### PR TITLE
Number literal object keys typechecking

### LINKED ISSUE #380: "non-string literal property keys not supported"
```
var x = {
    1234:"a",
    5678:"b"
}
```

Flow generates an error per integer key.

According to [11.1.5 of ECMA-262](http://www.ecma-international.org/ecma-262/5.1/#sec-11.1.5) this is a perfectly valid construct.

EDIT: @jgrund 's solution in https://github.com/facebook/flow/issues/380#issuecomment-224380551 works well

### PR DESCRIPTION
**Pull request by goodmind**
Fixes #380 

Doesn't allow in interfaces
Doesn't allow in class declarations

### CODE REVIEW

**Comment by goodmind** [comment]
Thanks @mroch for this library https://github.com/flowtype/ocaml-dtoa

**Comment by nmote** [comment]
Not sure who should review this. @dsainati1, @jbrown215, or @mvitousek, any of you have some time to spare for this?

**Comment by dsainati1** [comment]
Given that @mroch had raised concerns on #380 I'd like to get his opinion on this, but at the very least I think we'd want to see more tests here. It would be good to test number keys with different variance, as well as cases where objects are declared with number/string keys that are in conflict. I.e., what should happen here?
```
let a = {1 : "A", "1" : 3};
(a[1] : number)
```

**Comment by goodmind** [comment]
@dsainati1 it does the same as `{'1': 'A', '1': 3}` i.e. uses latter definition, added tests

**Comment by goodmind** [comment]
Parser test for contravariant number keys fails, probably it parses it as negative number

**Comment by goodmind** [comment]
Fixed contravariant properties parsing

**Comment by goodmind** [comment]
image
This probably should display numbers too

**Comment by goodmind** [comment]
@dsainati1 do you think tuples should be affected by this?
```js
//@flow

declare var e: { 0: number, 1: number, length: 2 };
declare var b: [number, number];

(e[0]: number);
(e[1]: number);

(b[0]: number);
(b[1]: number);

const m: { 0: number, 1: number, length: 2 } = b;
const m2: [number, number] = e;
```

File: src/typing/flow_js.ml [hunk 1]
```diff
@@ -5297,8 +5297,13 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
...
       let propref = match l with
       | DefT (reason_x, _, StrT (Literal (_, x))) ->
-          let reason_prop = replace_reason_const (RProperty (Some x)) reason_x in
-          Named (reason_prop, x)
+        let reason_prop = replace_reason_const (RProperty (Some x)) reason_x in
+        Named (reason_prop, x)
+      | DefT (reason_x, _, NumT (Literal (_, (x, _)))) ->
+        let x = Dtoa.ecma_string_of_float x in
+        let reason_prop = replace_reason_const (RProperty (Some x)) reason_x in
+        (* Probably should be different type constructor *)
```
Comment (goodmind) [inline src/typing/flow_js.ml]: Looks like this broke `{[key: number]: any}` reads,
```js
declare var e: { [key: number]: any };

e[0]; // string `0` [1] is incompatible with  number [2].Flow(InferError)
```

Comment (goodmind) [reply src/typing/flow_js.ml]: I think this https://github.com/facebook/flow/pull/2293#issuecomment-261344263 applies to this also, so probably I fixed it wrong

**Comment by goodmind** [comment]
This PR probably shouldn't mix number keys with string keys

**Comment by goodmind** [comment]
😃I forgot about this PR completely. I think to split parsing from typechecking in two PRs
@mroch do you have anything to say about number keys?

**Comment by goodmind** [comment]
Moved parsing here #7798

**Comment by nnmrts** [comment]
@goodmind Wow, thanks for your work on this! Do you have any updates for us?

**Comment by goodmind** [comment]
@nnmrts I don't think this would get in as-is, I hope at least parsing would get reviewed

**Comment by nnmrts** [comment]
@goodmind So it seems you're just getting ignored here. Nice! 👍👍👍

**Comment by goodmind** [comment]
@nnmrts not really, I have tons of PR merged already

**Comment by nnmrts** [comment]
Any update on this? 😆

**Comment by nnmrts** [comment]
Really? It's almost 9 months now!? This is ridiculous. Maintainers, then open up flowlib (and js types) for developers at least, if you don't want to fix major bugs like this one.

Flow is officially an unusable typechecker for 2014 javascript now. Numerical keys are standard since JUNE 2015!???

How is this even a typechecker when half of the types aren't following any standard?

**Pull request closed without merging**